File: aat-android/src/main/java/ch/bailu/aat/services/background/ProcessThread.java
package ch.bailu.aat.services.background;

import java.io.Closeable;

import ch.bailu.aat_lib.service.background.BackgroundTask;
import ch.bailu.aat_lib.service.background.ThreadControl;

public abstract class ProcessThread extends Thread implements Closeable, ThreadControl {

    private boolean continueThread=true;
    private final HandleStack queue;

    private BackgroundTask current = BackgroundTask.NULL;


    public ProcessThread(String name, HandleStack q) {
        super(name);
        queue = q;
        start();
    }


    public ProcessThread(String name, int limit) {
        this(name, new HandleStack(limit));
    }


    @Override
    public void run() {
        while(canContinue()) {

            try {
                current = queue.take();
                bgProcessHandle(current);
                current.onRemove();
                current = BackgroundTask.NULL;


            } catch (InterruptedException e) {
                continueThread=false;
                e.printStackTrace();
            }
        }
    }



    public abstract void bgOnHandleProcessed(BackgroundTask handle, long size);
    public abstract void bgProcessHandle(BackgroundTask handle);

    public void process(BackgroundTask handle) {
        if (canContinue()) queue.offer(handle);
    }


    @Override
    public void close() {
        continueThread=false;
        current.stopProcessing();
        queue.offer(BackgroundTask.STOP);
    }


    @Override
    public boolean canContinue() {
        return continueThread;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/service/background/ProcessThread.kt
package ch.bailu.aat_lib.service.background

import java.io.Closeable

abstract class ProcessThread(name: String, private val queue: HandleStack) : Thread(name), Closeable, ThreadControl {

    private var continueThread = true
    private var current = BackgroundTask.NULL

    init {
        start()
    }

    constructor(name: String, limit: Int) : this(name, HandleStack(limit))

    override fun run() {
        while (canContinue()) {
            try {
                current = queue.take()
                bgProcessHandle(current)
                current.onRemove()
                current = BackgroundTask.NULL
            } catch (e: InterruptedException) {
                continueThread = false
                e.printStackTrace()
            }
        }
    }

    abstract fun bgOnHandleProcessed(handle: BackgroundTask, size: Long)
    abstract fun bgProcessHandle(handle: BackgroundTask)

    fun process(handle: BackgroundTask) {
        if (canContinue()) queue.offer(handle)
    }

    override fun close() {
        continueThread = false
        current.stopProcessing()
        queue.offer(BackgroundTask.STOP)
    }

    override fun canContinue(): Boolean {
        return continueThread
    }
}
